### ISSUE TITLE
fancy-regex: problems with patterns in some syntaxes

### ISSUE THREAD

**Issue by sharkdp** [trishume/syntect#287]
I just tried the fancy-regex version of syntect for the first time (thank you for everyone involved!).

Here are a few of the problems I have encountered so far:

- The [PHP syntax in sublimehq/Packages](https://github.com/sublimehq/Packages/blob/759d6eed9b4beed87e602a23303a121c3a6c2fb3/PHP/PHP%20Source.sublime-syntax#L807) (master) uses some sort of inline comment:
  ```yaml
      - match: |-
        (?x)/\*\*(?:
          (?:\#@\+)?\s*$ (?# multi-line doc )
          |
          (?=\s+@.*\s\*/\s*$) (?# inline doc )
        )
  ```
  which causes the following error:
  ```
  Error while compiling regex '…': Unknown group flag
  ```
- The AsciiDoc syntax (https://github.com/SublimeText/AsciiDoc/blob/master/AsciiDoc.tmLanguage) fails with:
  ```
  Error while compiling regex '(?x)^
  (?= ([/+-.*_=]{4,})\s*(?m:$)
  | ([ \t]{1,})
  | [=]{1,6}\s*+
  | [ ]{,3}(?<marker>[-*_])([ ]{,2}\k<marker>){2,}[ \t]*+(?m:$)
  )': Unknown group flag
  ```
  (note that this is after `.sublime-syntax` conversion and possibly regex rewriting in syntect).
- The ARM Assembly syntax (https://github.com/tvi/Sublime-ARM-Assembly) failes with a similar error:
  ```
  Error while compiling regex '(?x)
  ^\s*\#\s*(define)\s+             # define
  ((?<id>[a-zA-Z_][a-zA-Z0-9_]*))  # macro name
  (?:                              # and optionally:
      (\()                         # an open parenthesis
          (
              \s* \g<id> \s*       # first argument
              ((,) \s* \g<id> \s*)*  # additional arguments
              (?:\.\.\.)?          # varargs ellipsis?
          )
      (\))                         # a close parenthesis
  )?': Unknown group flag
  ```
- The Haskell/Cabal syntax (https://github.com/SublimeHaskell/SublimeHaskell) fails with:
  ```
  Error while compiling regex '(=>|\u21D2)\s+([A-Z][\w']*)': Invalid escape
  ```
- Elixir (https://github.com/princemaple/elixir-sublime-syntax/) fails with a similar error:
  ```
  Error while compiling regex '(?x)
  \\g(?:
    <( ((?>-[1-9]\d*|\d+) | [a-zA-Z_][a-zA-Z_\d]{,31}) | \g<-1>?([^\[\\(){}|^$.?*+\n]+) )> | '\g<1>' |
    (<([^\[\\(){}|^$.?*+\n]+*)>? | '\g<-1>'?) )': Invalid escape
  ```
* Same for JavaScript/Babel (https://github.com/babel/babel-sublime):
  ```
  Error while compiling regex '(?x)
  (?:([_$a-zA-Z][$\w]*)\s*(=)\s*)?
  (?:\b(async)\s+)?
  (?=(\((?>(?>[^()]+)|\g<-1>)*\))\s*(=>))': Invalid escape
  ```
* The SLS syntax fails with
  ```
  Error while compiling regex '(?x)
  (?: ^ [ \t]* | [ \t]+ )
  (?:(\#) \p{Print}* )?
  (\n|\z)
  ': Regex error: regex parse error:
      (?:^[ \t]*|[ \t]+)(?:(\#)\p{Print}*)?(\n|\z)
                               ^^^^^^^^^
  error: Unicode property not found
  ```

This list continues for quite some time, but I'm not sure if it's worth to list them all. Most of them seem to be related to "unknown group flag" or "invalid escape".

Note: I just wanted to try this out within `bat`, there is absolutely no "pressure" to get this fixed (as always, of course :smile:). I was just curious and thought this might help.

**Comment by trishume** [comment]
cc @robinst any thoughts on what might needed to be added to either the rewriter or fancy-regex to fix these and guesses as to how much work it would be?

**Comment by Keats** [comment]
Ah damn I was about to try it in Zola and rust-onig still hasn't shipped a new version with optional bindgen :(

@sharkdp how do you feel about creating a repository to contain all .sublime-syntaxes? We are kind of duplicating work in Zola and bat (and potentially other tools) having our own repo of syntaxes.

**Comment by sharkdp** [comment]
> @sharkdp how do you feel about creating a repository to contain all .sublime-syntaxes? We are kind of duplicating work in Zola and bat (and potentially other tools) having our own repo of syntaxes.

Sounds like a great idea.  Should we move this discussion to `bat`s issue tracker? There are probably a lot of details which would need to be figured out (what to include? what *not* to include? how to deal with temporary patches, as `bat` currently does? what kind of tooling do we want to share? etc.)

**Comment by robinst** [comment]
> cc @robinst any thoughts on what might needed to be added to either the rewriter or fancy-regex to fix these and guesses as to how much work it would be?

Yeah, I think some of these should be easy to fix in fancy-regex, e.g. `(?# ...)` and `\u`, I'll work on those first. Others are a bit more work like named capture groups, but on the radar: https://github.com/fancy-regex/fancy-regex/issues/34

**Comment by Keats** [comment]
@robinst any chance you can release a new version of fancy-regex? I want to update some syntaxes and want to know exactly which ones are failing after your latest patches.

**Comment by robinst** [comment]
@Keats Published version 0.3.4 now: https://github.com/fancy-regex/fancy-regex/blob/master/CHANGELOG.md#034---2020-04-28

Note that it's unlikely that I'll implement `\g<...>` for [subexp calls](https://github.com/kkos/oniguruma/blob/master/doc/RE#L451) in the near future. I would recommend replacing them with `{{variable}}` references instead. Note that that might even make it better for Sublime Text itself, as I don't think [sregex](https://github.com/openresty/sregex/) implements that syntax either.

**Comment by robinst** [comment]
Also created a PR to make error messages include what the unknown flag/escape is: https://github.com/fancy-regex/fancy-regex/pull/46

**Comment by Keats** [comment]
Thanks a lot @robinst , I'll give it a try asap

**Comment by robinst** [comment]
Released https://github.com/fancy-regex/fancy-regex/pull/46 as 0.3.5

**Comment by Keats** [comment]
It looks like a lot of syntaxes are using `\G` (TypeScript like https://github.com/getzola/zola/blob/bc496e61010be1094a9192003ea59506c14d9397/sublime/syntaxes/TypeScript.sublime-syntax#L518) and `\g` (Elixir Regex https://github.com/princemaple/elixir-sublime-syntax/blob/master/Regular%20Expressions%20(Elixir).sublime-syntax#L52 and probably a few others). I had never heard of subexp calls before, could the linked Elixir file be changed to not use `\g`?

**Comment by robinst** [comment]
`\G` is an interesting one, it will anchor the match at where the current search begins. I wonder if that's necessary for correctness for those syntaxes or if they do it for performance (but not sure it would actually make things better). @keith-hall maybe you have some insight for that?

For `\g`, it should be possible to replace by repeating the pattern or using variables. E.g. this:

```
<({{capture_name}})>|'\g<-1>'|\g<-1>
```

Should be the same as:

```
<({{capture_name}})>|'({{capture_name}})'|({{capture_name}})
```

(But `captures` might need to be extended.)

**Comment by keith-hall** [comment]
`\G` is needed for correctness - this was often used in `tmLanguage` grammars where there was less functionality for working with contexts. Pretty sure it wouldn't have much impact on performance, or it would slow if down if no matches were found and all the patterns in the context use `\G`, as it would move the string pointer one char along and try again until it finds a match.

**Comment by robinst** [comment]
Hmm so do you know why none of the built-in syntaxes use `\G` at all? Seems strange. I wonder if sregex doesn't support it either and so they avoid it.

**Comment by keith-hall** [comment]
I believe that is correct, yes.

**Comment by robinst** [comment]
fancy-regex 0.4.0 now supports named groups, see [changelog](https://github.com/fancy-regex/fancy-regex/blob/main/CHANGELOG.md#040---2020-09-27). `\g` and `\G` are not yet supported.

**Comment by keith-hall** [comment]
Subroutine support in fancy-regex is nearly ready: https://github.com/fancy-regex/fancy-regex/pull/230

Also `\G` has been supported for a long time now

**Comment by tvanderstad** [comment]
Does that mean this issue is unblocked? I'm using syntect to build an iOS app, where `onig-sys` is difficult to ship. `fancy-regex` is easy but as a result of this issue there's a list of languages my app doesn't support. I would be pleased to even just add JavaScript support.

**Comment by keith-hall** [comment]
Yep, https://github.com/fancy-regex/fancy-regex/releases/tag/0.18.0 has very good Oniguruma compatibility. (A syntax highlighter using lots of textmate grammars no longer experiences any regex compilation errors.) If you still see issues with that version, please report them on the fancy-regex issue tracker.
